src/components/ConfigEditor/AzureCredentialsConfig.ts

```diff
@@ -0,0 +1,197 @@
+import { DataSourceSettings } from '@grafana/data';
+import { config } from '@grafana/runtime';
+
+import { AzureCloud, AzureCredentials, ConcealedSecret } from './AzureCredentials';
+
+const concealed: ConcealedSecret = Symbol('Concealed client secret');
+
+export const getOboEnabled = (): boolean => !!config.featureToggles['adxOnBehalfOf'];
+
+function getDefaultAzureCloud(): string {
+  return config.azure.cloud || AzureCloud.Public;
+}
+
+function getSecret(options: DataSourceSettings<any, any>): undefined | string | ConcealedSecret {
+  if (options.secureJsonFields.azureClientSecret) {
+    // The secret is concealed on server
+    return concealed;
+  } else {
+    const secret = options.secureJsonData?.azureClientSecret;
+    return typeof secret === 'string' && secret.length > 0 ? secret : undefined;
+  }
+}
+
+export function getDefaultCredentials(): AzureCredentials {
+  if (config.azure.managedIdentityEnabled) {
+    return { authType: 'msi' };
+  } else {
+    return { authType: 'clientsecret', azureCloud: getDefaultAzureCloud() };
+  }
+}
+
+export function getCredentials(options: DataSourceSettings<any, any>): AzureCredentials {
+  const credentials = options.jsonData.azureCredentials as AzureCredentials | undefined;
+
+  // If no credentials saved then try to restore legacy credentials, otherwise return default credentials
+  if (!credentials) {
+    return getLegacyCredentials(options) ?? getDefaultCredentials();
+  }
+
+  switch (credentials.authType) {
+    case 'msi':
+      if (config.azure.managedIdentityEnabled) {
+        return {
+          authType: 'msi',
+        };
+      } else {
+        // If authentication type is managed identity but managed identities were disabled in Grafana config,
+        // then we should fall back to an empty default credentials
+        return getDefaultCredentials();
+      }
+    case 'clientsecret':
+    case 'clientsecret-obo':
+      if (credentials.authType === 'clientsecret-obo' && !getOboEnabled()) {
+        // If authentication type is OBO but OBO were disabled in Grafana config,
+        // then we should fall back to an empty default credentials
+        return getDefaultCredentials();
+      }
+      return {
+        authType: credentials.authType,
+        azureCloud: credentials.azureCloud || getDefaultAzureCloud(),
+        tenantId: credentials.tenantId,
+        clientId: credentials.clientId,
+        clientSecret: getSecret(options),
+      };
+  }
+}
+
+function getLegacyCredentials(options: DataSourceSettings<any, any>): AzureCredentials | undefined {
+  const jsonData = options.jsonData;
+
+  if (typeof jsonData.tenantId === 'string' || typeof jsonData.clientId === 'string') {
+    try {
+      const azureCloud = resolveLegacyCloudName(jsonData.cloudName);
+
+      return {
+        authType: jsonData.onBehalfOf ? 'clientsecret-obo' : 'clientsecret',
+        azureCloud: azureCloud,
+        tenantId: typeof jsonData.tenantId === 'string' ? jsonData.tenantId : '',
```

Comment:
This type and `clientId` below shouldn't need to be re-validated.

Reply:
The type needs to be re-evaluated because the initial condition is `OR`, it doesn't guarantee that both fields are string:

```go
typeof jsonData.tenantId === 'string' || typeof jsonData.clientId === 'string'
```

I use `OR` because even if the existing credentials are incomplete and some input is missing, I still attempt to migrate whichever is present, as opposed to having everything blank because some fields were missing.